MigrationTutorial/Migrations/V3Utils.cs

```diff
@@ -62,58 +67,58 @@ public static void SeedData()
+        {
+            // it's assumed that there's always 1 and 1 only type of Consumable
+            var oldConsumable = ((IQueryable<RealmObject>)migration.OldRealm.DynamicApi.All("Consumable")).Filter("_Type == $0", consumableType).FirstOrDefault();
+            if (oldConsumable == null)
             {
-                var oldConsumable = oldConsumables.ElementAt(i);
-                if (oldConsumable._Type == "GlueHolder")
-                {
-                    var supplierId = (ObjectId)oldConsumable.Supplier.Id;
-                    var glueSupplier = migration.NewRealm.All<Supplier>().Filter("Id == $0", supplierId).First();
-
-                    migration.NewRealm.Add(new MachineryAndTool()
-                    {
-                        Type = Type.ManufacturingTool,
-                        Status = OperationalStatus.Functioning,
-                        AssignedMaintaner = null,
-                        ToolName = oldConsumable._Type,
-                        Supplier = glueSupplier,
-                        Brand = oldConsumable.Brand
-                    });
-
-                    oldGlueHolderId = oldConsumable.ProductId;
-                }
-                else if (oldConsumable._Type == "Brush")
-                {
-                    var supplierId = (ObjectId)oldConsumable.Supplier.Id;
-                    var brushSupplier = migration.NewRealm.All<Supplier>().Filter("Id == $0", supplierId).First();
+                Logger.LogWarning($"No consumable was found with type {consumableType}. Nothing to convert.");
+                return;
+            }
 
-                    migration.NewRealm.Add(new MachineryAndTool()
-                    {
-                        Type = Type.ManufacturingTool,
-                        Status = OperationalStatus.Functioning,
-                        AssignedMaintaner = null,
-                        ToolName = oldConsumable._Type,
-                        Supplier = brushSupplier,
-                        Brand = oldConsumable.Brand
-                    });
+            Supplier consumableSupplier = null;
+            string consumableBrand = string.Empty;
 
-                    oldBrushId = oldConsumable.ProductId;
-                }
+            try
+            {
+                var supplierId = oldConsumable.DynamicApi.Get<RealmObject>("Supplier").DynamicApi.Get<ObjectId>("Id");
+                consumableSupplier = migration.NewRealm.All<Supplier>().Filter("Id == $0", supplierId).FirstOrDefault();
+                consumableBrand = oldConsumable.DynamicApi.Get<string>("Brand");
+            }
+            catch (System.MissingMemberException)
+            {
+                Logger.LogDebug($"Some properties don't exist on the old realm. This could likely mean that a migration from schema V1 to schema V3 was performed, not passing through 2.\n The operation will continue leaving such fields at their default value.");
```

Comment:
Not sure I completely understand but isn't the tutorial running through this v2 every time? Or isn't that the case anymore with the new structure (pre-compiler statements)?

Reply:
It's now changed in a way that each migration is self contained. And this is all possible because of the preprocessor directives.

Reply:
👍